// @tapjs/.test-built/dist/mjs/index.js
// This file is automatically generated, please do not edit
import { parseTestArgs, TestBase, } from '@tapjs/core';
const copyToString = (v) => ({
    toString: Object.assign(() => v.toString(), {
        toString: () => 'function toString() { [native code] }',
    }),
});
import _Plugin_stdin from "@tapjs/core/plugin/stdin.js";
import _Plugin_spawn from "@tapjs/core/plugin/spawn.js";
import _Plugin_beforeEach from "@tapjs/core/plugin/before-each.js";
import _Plugin_afterEach from "@tapjs/core/plugin/after-each.js";
import _Plugin_asserts from "@tapjs/asserts";
export const Plugin_stdin = _Plugin_stdin;
export const Plugin_spawn = _Plugin_spawn;
export const Plugin_beforeEach = _Plugin_beforeEach;
export const Plugin_afterEach = _Plugin_afterEach;
export const Plugin_asserts = _Plugin_asserts;
const plugins = [
    Plugin_stdin,
    Plugin_spawn,
    Plugin_beforeEach,
    Plugin_afterEach,
    Plugin_asserts,
];
const applyPlugins = (base, ext = [
    ...plugins.map(p => p(base, base.options)),
    base,
]) => {
    const getCache = new Map();
    const t = new Proxy(base, {
        has(_, p) {
            for (const t of ext) {
                if (Reflect.has(t, p))
                    return true;
            }
            return false;
        },
        ownKeys() {
            const k = [];
            for (const t of ext) {
                const keys = Reflect.ownKeys(t);
                k.push(...keys);
            }
            return [...new Set(k)];
        },
        getOwnPropertyDescriptor(_, p) {
            for (const t of ext) {
                const prop = Reflect.getOwnPropertyDescriptor(t, p);
                if (prop)
                    return prop;
            }
            return undefined;
        },
        set(_, p, v) {
            // check to see if there's any setters, and if so, set it there
            // otherwise, just set on the base
            for (const t of ext) {
                let o = t;
                while (o) {
                    if (Reflect.getOwnPropertyDescriptor(o, p)?.set) {
                        //@ts-ignore
                        t[p] = v;
                        return true;
                    }
                    o = Reflect.getPrototypeOf(o);
                }
            }
            //@ts-ignore
            base[p] = v;
            return true;
        },
        get(_, p) {
            // cache get results so t.blah === t.blah
            // we only cache functions, so that getters aren't memoized
            // Of course, a getter that returns a function will be broken,
            // at least when accessed from outside the plugin, but that's
            // a pretty narrow caveat, and easily documented.
            if (getCache.has(p))
                return getCache.get(p);
            for (const plug of ext) {
                if (p in plug) {
                    //@ts-ignore
                    const v = plug[p];
                    // Functions need special handling so that they report
                    // the correct toString and are called on the correct object
                    // Otherwise attempting to access #private props will fail.
                    if (typeof v === 'function') {
                        const f = function (...args) {
                            const thisArg = this === t ? plug : this;
                            return v.apply(thisArg, args);
                        };
                        const vv = Object.assign(f, copyToString(v));
                        const nameProp = Reflect.getOwnPropertyDescriptor(v, 'name');
                        if (nameProp) {
                            Reflect.defineProperty(f, 'name', nameProp);
                        }
                        getCache.set(p, vv);
                        return vv;
                    }
                    else {
                        getCache.set(p, v);
                        return v;
                    }
                }
            }
        },
    });
    Object.assign(base, { t });
    ext.unshift({ t });
    return t;
};
export class Test extends TestBase {
    constructor(opts) {
        super(opts);
        return applyPlugins(this);
    }
    test(...args) {
        const extra = parseTestArgs(...args);
        extra.todo = true;
        const Class = this.constructor;
        return this.sub(Class, extra, this.test);
    }
    todo(...args) {
        const extra = parseTestArgs(...args);
        extra.todo = true;
        const Class = this.constructor;
        return this.sub(Class, extra, this.test);
    }
    skip(...args) {
        const extra = parseTestArgs(...args);
        extra.skip = true;
        const Class = this.constructor;
        return this.sub(Class, extra, this.test);
    }
}
