// ink/build/render.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _instance = _interopRequireDefault(require("./instance"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const instances = new WeakMap();

var _default = (node, options = {}) => {
  // Stream was passed instead of `options` object
  if (typeof options.write === 'function') {
    options = {
      stdout: options,
      stdin: process.stdin
    };
  }

  options = _objectSpread({
    stdout: process.stdout,
    stdin: process.stdin,
    debug: false,
    exitOnCtrlC: true
  }, options);
  let instance;

  if (instances.has(options.stdout)) {
    instance = instances.get(options.stdout);
  } else {
    instance = new _instance.default(options);
    instances.set(options.stdout, instance);
  }

  instance.render(node);
  return {
    rerender: instance.render,
    unmount: instance.unmount,
    waitUntilExit: instance.waitUntilExit,
    cleanup: () => instances.delete(options.stdout)
  };
};

exports.default = _default;
