// @tapjs/.test-built/dist/cjs/index.js
"use strict";
// This file is automatically generated, please do not edit
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Test = exports.pluginsLoaded = exports.Plugin_filter = exports.Plugin_intercept = exports.Plugin_mock = exports.Plugin_fixture = exports.Plugin_snapshot = exports.Plugin_asserts = exports.Plugin_afterEach = exports.Plugin_after = exports.Plugin_beforeEach = exports.Plugin_before = exports.Plugin_spawn = exports.Plugin_stdin = void 0;
const core_1 = require("@tapjs/core");
const copyToString = (v) => ({
    toString: Object.assign(() => v.toString(), {
        toString: () => 'function toString() { [native code] }',
    }),
});
const copyFunction = (t, plug, v) => {
    const f = function (...args) {
        const thisArg = this === t ? plug : this;
        return v.apply(thisArg, args);
    };
    const vv = Object.assign(f, copyToString(v));
    const nameProp = Reflect.getOwnPropertyDescriptor(v, 'name');
    if (nameProp) {
        Reflect.defineProperty(f, 'name', nameProp);
    }
    return vv;
};
const stdin_1 = __importDefault(require("@tapjs/core/plugin/stdin"));
const spawn_1 = __importDefault(require("@tapjs/core/plugin/spawn"));
const before_1 = __importDefault(require("@tapjs/core/plugin/before"));
const before_each_1 = __importDefault(require("@tapjs/core/plugin/before-each"));
const after_1 = __importDefault(require("@tapjs/core/plugin/after"));
const after_each_1 = __importDefault(require("@tapjs/core/plugin/after-each"));
const asserts_1 = __importDefault(require("@tapjs/asserts"));
const snapshot_1 = __importDefault(require("@tapjs/snapshot"));
const fixture_1 = __importDefault(require("@tapjs/fixture"));
const mock_1 = __importDefault(require("@tapjs/mock"));
const intercept_1 = __importDefault(require("@tapjs/intercept"));
const filter_1 = __importDefault(require("@tapjs/filter"));
exports.Plugin_stdin = stdin_1.default;
exports.Plugin_spawn = spawn_1.default;
exports.Plugin_before = before_1.default;
exports.Plugin_beforeEach = before_each_1.default;
exports.Plugin_after = after_1.default;
exports.Plugin_afterEach = after_each_1.default;
exports.Plugin_asserts = asserts_1.default;
exports.Plugin_snapshot = snapshot_1.default;
exports.Plugin_fixture = fixture_1.default;
exports.Plugin_mock = mock_1.default;
exports.Plugin_intercept = intercept_1.default;
exports.Plugin_filter = filter_1.default;
const plugins = [
    exports.Plugin_stdin,
    exports.Plugin_spawn,
    exports.Plugin_before,
    exports.Plugin_beforeEach,
    exports.Plugin_after,
    exports.Plugin_afterEach,
    exports.Plugin_asserts,
    exports.Plugin_snapshot,
    exports.Plugin_fixture,
    exports.Plugin_mock,
    exports.Plugin_intercept,
    exports.Plugin_filter,
];
exports.pluginsLoaded = new Map([
    ['stdin', exports.Plugin_stdin],
    ['spawn', exports.Plugin_spawn],
    ['before', exports.Plugin_before],
    ['beforeEach', exports.Plugin_beforeEach],
    ['after', exports.Plugin_after],
    ['afterEach', exports.Plugin_afterEach],
    ['asserts', exports.Plugin_asserts],
    ['snapshot', exports.Plugin_snapshot],
    ['fixture', exports.Plugin_fixture],
    ['mock', exports.Plugin_mock],
    ['intercept', exports.Plugin_intercept],
    ['filter', exports.Plugin_filter],
]);
const applyPlugins = (base, ext = [
    ...plugins.map(p => p(base, base.options)),
    base,
]) => {
    const getCache = new Map();
    // extend the proxy with Object.create, and then set the toStringTag
    // to 'Test', so we don't get stack frames like `Proxy.<anonymous>`
    const t = Object.create(new Proxy(base, {
        has(_, p) {
            for (const t of ext) {
                if (Reflect.has(t, p))
                    return true;
            }
            return false;
        },
        ownKeys() {
            const k = [];
            for (const t of ext) {
                const keys = Reflect.ownKeys(t);
                k.push(...keys);
            }
            return [...new Set(k)];
        },
        getOwnPropertyDescriptor(_, p) {
            for (const t of ext) {
                const prop = Reflect.getOwnPropertyDescriptor(t, p);
                if (prop)
                    return prop;
            }
            return undefined;
        },
        set(_, p, v) {
            // check to see if there's any setters, and if so, set it there
            // otherwise, just set on the base
            let didSet = false;
            for (const t of ext) {
                let o = t;
                while (o) {
                    // assign to the all plugs that can receive it
                    const prop = Reflect.getOwnPropertyDescriptor(o, p);
                    if (prop) {
                        if (prop.set || prop.writable) {
                            //@ts-ignore
                            t[p] = v;
                            didSet = true;
                        }
                        break;
                    }
                    o = Reflect.getPrototypeOf(o);
                }
            }
            if (!didSet) {
                // if nothing has that field, assign to the base
                //@ts-ignore
                base[p] = v;
            }
            return true;
        },
        get(_, p) {
            if (p === Symbol.toStringTag)
                return 'Test';
            if (p === 'parent') {
                return base.parent?.t;
            }
            // cache get results so t.blah === t.blah
            // we only cache functions, so that getters aren't memoized
            // Of course, a getter that returns a function will be broken,
            // at least when accessed from outside the plugin, but that's
            // a pretty narrow caveat, and easily documented.
            if (getCache.has(p))
                return getCache.get(p);
            for (const plug of ext) {
                if (p in plug) {
                    //@ts-ignore
                    const v = plug[p];
                    // Functions need special handling so that they report
                    // the correct toString and are called on the correct object
                    // Otherwise attempting to access #private props will fail.
                    if (typeof v === 'function') {
                        const vv = copyFunction(t, plug, v);
                        getCache.set(p, vv);
                        return vv;
                    }
                    else if (p === 'parent') {
                        return v?.t;
                    }
                    else {
                        return v;
                    }
                }
            }
        },
    }));
    Object.defineProperty(t, Symbol.toStringTag, {
        value: 'Test',
    });
    Object.assign(base, { t });
    ext.unshift({ t });
    return t;
};
class Test extends core_1.TestBase {
    constructor(opts) {
        super(opts);
        return applyPlugins(this);
    }
    static get plugins() {
        return exports.pluginsLoaded;
    }
    static pluginLoaded(plugin) {
        return plugins.includes(plugin);
    }
    pluginLoaded(plugin) {
        return Test.pluginLoaded(plugin);
    }
    get plugins() {
        return Test.plugins;
    }
    test(...args) {
        const extra = (0, core_1.parseTestArgs)(...args);
        return this.sub(Test, extra, this.test);
    }
    todo(...args) {
        const extra = (0, core_1.parseTestArgs)(...args);
        extra.todo = true;
        return this.sub(Test, extra, this.todo);
    }
    skip(...args) {
        const extra = (0, core_1.parseTestArgs)(...args);
        extra.skip = true;
        return this.sub(Test, extra, this.skip);
    }
}
exports.Test = Test;
